src/main/scala/viper/gobra/translator/context/Context.scala

```diff
@@ -0,0 +1,190 @@
+
+import viper.gobra.ast.internal._
+import viper.gobra.translator.encodings.combinators.{DefaultEncoding, TypeEncoding}
+import viper.gobra.translator.util.ViperWriter.{CodeWriter, MemberWriter}
+import viper.gobra.ast.{internal => in}
+import viper.gobra.translator.library.arrays.Arrays
+import viper.gobra.translator.library.conditions.Conditions
+import viper.gobra.translator.library.equality.Equality
+import viper.gobra.translator.library.fields.Fields
+import viper.gobra.translator.library.fixpoints.Fixpoint
+import viper.gobra.translator.library.multiplicity.SeqMultiplicity
+import viper.gobra.translator.library.options.Options
+import viper.gobra.translator.library.slices.Slices
+import viper.gobra.translator.library.tos.{OptionToSeq, SeqToMultiset, SeqToSet}
+import viper.gobra.translator.library.tuples.Tuples
+import viper.gobra.translator.library.unknowns.UnknownValues
+import viper.silver.{ast => vpr}
+
+trait Context {
+
+  // components
+  def field: Fields
+
+  def array: Arrays
+
+  def seqToSet: SeqToSet
+
+  def seqToMultiset: SeqToMultiset
+
+  def seqMultiplicity: SeqMultiplicity
+
+  def option: Options
+
+  def optionToSeq: OptionToSeq
+
+  def slice: Slices
+
+  def fixpoint: Fixpoint
+
+  def tuple: Tuples
+
+  def equality: Equality
+
+  def condition: Conditions
+
+  def unknownValue: UnknownValues
+
+  // translator
+  protected def typeEncoding: TypeEncoding
+
+  def defaultEncoding: DefaultEncoding
+
+  def typ(x: in.Type): vpr.Type = typeEncoding.typ(this)(x)
+
+  def variable(x: in.BodyVar): vpr.LocalVarDecl = typeEncoding.variable(this)(x)
+
+  def globalVar(x: in.GlobalVar): CodeWriter[vpr.Exp] = typeEncoding.globalVar(this)(x)
+
+  def member(x: in.Member): MemberWriter[Vector[vpr.Member]] = typeEncoding.member(this)(x)
+
+  def method(x: in.Member): MemberWriter[vpr.Method] = typeEncoding.finalMethod(this)(x)
+
+  def function(x: in.Member): MemberWriter[vpr.Function] = typeEncoding.finalFunction(this)(x)
+
+  def predicate(x: in.Member): MemberWriter[vpr.Predicate] = typeEncoding.finalPredicate(this)(x)
+
+  def varPrecondition(x: in.Parameter.In): Option[MemberWriter[vpr.Exp]] = typeEncoding.varPrecondition(this).lift(x)
+
+  def varPostcondition(x: in.Parameter.Out): Option[MemberWriter[vpr.Exp]] = typeEncoding.varPostcondition(this).lift(x)
+
+  def initialization(x: in.Location): CodeWriter[vpr.Stmt] = typeEncoding.initialization(this)(x)
+
+  def assignment(x: in.Assignee, rhs: in.Expr)(src: in.Node): CodeWriter[vpr.Stmt] = typeEncoding.assignment(this)(x, rhs, src)
+
+  def equal(lhs: in.Expr, rhs: in.Expr)(src: in.Node): CodeWriter[vpr.Exp] = typeEncoding.equal(this)(lhs, rhs, src)
+
+  def goEqual(lhs: in.Expr, rhs: in.Expr)(src: in.Node): CodeWriter[vpr.Exp] = typeEncoding.goEqual(this)(lhs, rhs, src)
+
+  def expr(x: in.Expr): CodeWriter[vpr.Exp] = typeEncoding.finalExpression(this)(x)
+
+  def ass(x: in.Assertion): CodeWriter[vpr.Exp] = typeEncoding.finalAssertion(this)(x)
```

Comment:
I used long names for the typeencoding functions and shorter names for the context functions.

Reply:
In context, for some of the functions I was unsure if I should use shorter names. Some ideas are:

- `variable` -> `var`
- `initialization` -> `init`
- `assignment` -> `assign`
- I think `equal` -> `eq` would cause issues with the native eq method.
- `invariant` -> `inv`
- `precondition` -> `pre` (not sure about this one)
- `postcondition` -> `post` (again, not sure about this one)